### PR TITLE
Instance pool usages

### PR DESCRIPTION
**Pull request by johnpaulkee**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by johnpaulkee** [comment]
I plan to add instancePoolUsages.json and instancePools.json (not in this review) in a separate PR, so merging with just these files after validation and review is ok.

**Comment by ravbhatnagar** [comment]
@johnpaulkee - The API signature and the response schema does not comply with ARM /usages contract. Please take a look - https://armwiki.azurewebsites.net/api_contracts/UsagesAPIContract.html

**Comment by ravbhatnagar** [comment]
We will need to update the usages contract to cater to such scenarios. Will have a proposal early next week. That way we can have consistency across Azure w.r.t getting usage data.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        },
+        "requestedLimit": {
+          "format": "int32",
+          "description": "The requested limit.",
+          "type": "integer",
+          "readOnly": true
+        }
+      }
+    },
+    "InstancePoolUsage": {
```
Comment (sanjaiganesh) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: >InstancePoolUsage [](start = 5, length = 17)

and per spec, name is not just a string.. it is a json object.

"name": {
        "value": "Storage",
        "localizedValue": "Storage"
      }

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+          "nextLinkName": "nextLink"
+        },
+        "x-ms-examples": {
+          "List instance pool usages.": {
+            "$ref": "./examples/InstancePoolUsageListByInstancePool.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/instancePools/{instancePoolName}/usages/{usageName}": {
```
Comment (sanjaiganesh) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: >{usageName} [](start = 136, length = 11)

also per spec, usageName is part of query parameter.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+          "type": "integer",
+          "readOnly": true
+        },
+        "currentLimit": {
+          "format": "int32",
+          "description": "Instance pool current limit.",
+          "type": "integer",
+          "readOnly": true
+        },
+        "requestedLimit": {
```
Comment (sanjaiganesh) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: >requestedLimit [](start = 9, length = 14)

this is just "limit" according to the spec.. is this PR updated based on the usages api spec shared offline ?.. for ex: unit doesn't allow VCores etc..

**Comment by sanjaiganesh** [comment]
PR still needs to be updated per new usages spec (ARM). Gaurav is following up offline with the team.

---
In reply to: [493736164](https://github.com/Azure/azure-rest-api-specs/pull/5894#issuecomment-493736164) [](ancestors = 493736164)

**Comment by johnpaulkee** [comment]
Acked on updated usages contract, also acked on usages name being a query param.
Pasting updated contract with usages changes below.

```
GET /instancePool/instancePool1/usages
GET /subscriptions/{subscriptionId}/providers/{providerNamespace}/usages?scope={ subscription/…/…/instancePool/instancePool1 }
 
Value: [{ 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/vcore_utilization”, 
      "name": { 
        "value": "VCore utilization", 
        "localizedValue": "VCore utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "VCores", 
      "currentValue": 12, 
      "limit": 16, 
      "requestedLimit": 40, 
    },
    { 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/storage_utilization”, 
      "name": { 
        "value": "Storage utilization", 
        "localizedValue": "Storage utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "Gigabytes", 
      "currentValue": 256, 
      "limit": 8196, 
      "requestedLimit": null,
    },
    {
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/database_utilization”, 
      "name": { 
        "value": "Database utilization", 
        "localizedValue": "Database utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "Number of Databases", 
      "currentValue": 53, 
      "limit": 100, 
      "requestedLimit": null, 
    }
]
 
 
These would also give equivalent responses:
GET /instancePool/ipName/usages?expandChildren=true
GET /subscriptions/{subscriptionId}/providers/{providerNamespace}/usages?scope={ subscription/…/…/instancePool/instancePool1 }&expandChildren=true
 
We care about id in the collection call below so that we can apply RBAC access.
 
Value: [{ 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/vcore_utilization”, 
      "name": { 
        "value": "VCore utilization", 
        "localizedValue": "VCore utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "VCores", 
      "currentValue": 12, 
      "limit": 16, 
      "requestedLimit": 40, 
    },
    { 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/storage_utilization”, 
      "name": { 
        "value": "Storage utilization", 
        "localizedValue": "Storage utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "Gigabytes", 
      "currentValue": 256, 
      "limit": 8196, 
      "requestedLimit": null,
    },
    {
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/usages/database_utilization”, 
      "name": { 
        "value": "Database utilization", 
        "localizedValue": "Database utilization" 
      }, 
      "type": "Microsoft.Sql/instancePools/usages", 
      "unit": "Number of Databases", 
      "currentValue": 53, 
      "limit": 100, 
      "requestedLimit": null, 
    },
    { 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/managedInstances/managedInstance1/usages”,
      "name": {
        "value": "VCore utilization", 
        "localizedValue": "VCore utilization" 
      }, 
      "type": "Microsoft.Sql/managedInstances/usages", 
      "unit": "Gigabytes", 
      "currentValue": null, 
      "limit": 256, 
      "requestedLimit": 4096,
    },
    { 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/managedInstances/managedInstance1/usages”,
      "name": {
        "value": "Storage utilization", 
        "localizedValue": "Storage utilization" 
      }, 
      "type": "Microsoft.Sql/managedInstances/usages", 
      "unit": "Gigabytes", 
      "currentValue": null, 
      "limit": 256, 
      "requestedLimit": 4096,
    },
    { 
      "id": "/subscriptions/…/resourceGroups/…/providers/Microsoft.Sql/instancePools/instancePool1/managedInstances/managedInstance1/usages”,
      "name": { 
        "value": "Database utilization", 
        "localizedValue": "Database utilization" 
      }, 
      "type": "Microsoft.Sql/managedInstances/usages", 
      "unit": "Number of databases", 
      "currentValue": 15,
      "limit": 100, 
      "requestedLimit": null,
    },
    …
]
 

```

**Comment by johnpaulkee** [comment]
> PR still needs to be updated per new usages spec (ARM). Gaurav is following up offline with the team.
> 
> In reply to: [493736164](https://github.com/Azure/azure-rest-api-specs/pull/5894#issuecomment-493736164) [](ancestors = 493736164)

I've updated swagger based on updated usages contract.
There is ongoing thread with ARM team to help us with this:

```
-	Provide exact standardized swagger, in particular:
o	prescribed operation id’s (e.g. “InstancePools_ListUsages”? “Usages_ListByInstancePool”),
o	schema names
o	parameter names 
o	descriptions,
-	Work with Samer’s team to come up with validators for these?
```

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json (hunk 3 above)
Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: updated to limit.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json (hunk 2 above)
Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: acked. removed need for usageName. will just call /usages

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json (hunk 1 above)
Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: updated name to json object.

**Comment by sergey-shandar** [comment]
@ravbhatnagar @sanjaiganesh any update on ARM review?

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+          "readOnly": true
+        },
+        "nextLink": {
+          "description": "Link to retrieve next page of results.",
+          "type": "string",
+          "readOnly": true
+        }
+      }
+    },
+    "Usage": {
```
Comment (chidmdxx) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: do you need to add quotaPeriod and nextResetTime? going from the design doc

Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: they were optional I believe.

Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: image

Yeah they're not required.

Comment (chidmdxx) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: ok, as I said wasn't sure you needed them just wanted to be thorough

---
In reply to: [293155183](https://github.com/Azure/azure-rest-api-specs/pull/5894#discussion_r293155183) [](ancestors = 293155183)

Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: Sounds good :) appreciate it.

**Comment by johnpaulkee** [comment]
Ping @ravbhatnagar  for review.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json [hunk 5]
```diff
@@ -0,0 +1,251 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2.0",
```
Comment (ravbhatnagar) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: this should be the api-version

**Comment by ravbhatnagar** [comment]
@johnpaulkee - looks good. just the api-version needs to be fixed. Signing off considering the api-version value will be fixed.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json (hunk 5 above)
Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/instancePoolUsages.json]: fixed. thanks @ravbhatnagar 👍

**Comment by johnpaulkee** [comment]
Thanks @ravbhatnagar , I've updated api-version.
@sergey-shandar should be good for review now.

**Comment by johnpaulkee** [comment]
ping @sergey-shandar  for review.
I've updated the instancePoolUsages.json to usages.json FYI.

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/usages.json [hunk 6]
```diff
@@ -0,0 +1,251 @@
+{
```
Comment (sergey-shandar) [inline specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/usages.json]: This new file is not referenced from a readme.md file. https://dev.azure.com/azure-sdk/public/_build/results?buildId=51672&view=logs

**Review by sergey-shandar — CHANGES_REQUESTED** [review]

File: specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/usages.json (hunk 6 above)
Comment (johnpaulkee) [reply specification/sql/resource-manager/Microsoft.Sql/preview/2018-06-01-preview/usages.json]: updated to have readme.md reference usages.json

**Review by sergey-shandar — APPROVED** [review]

**Comment by johnpaulkee** [comment]
Thanks @sergey-shandar !

**Pull request merged**